command/ca/certificate.go

```diff
@@ -131,40 +132,54 @@ func certificateAction(ctx *cli.Context) error {
 		return err
 	}
 
-	var isStepToken bool
-	if len(token) == 0 {
-		if token, err = flow.GenerateToken(ctx, subject, sans); err != nil {
+	if len(tok) == 0 {
+		if tok, err = flow.GenerateToken(ctx, subject, sans); err != nil {
 			return err
 		}
-		isStepToken = isStepCertificatesToken(token)
-	} else {
-		isStepToken = isStepCertificatesToken(token)
-		if isStepToken && len(sans) > 0 {
-			return errs.MutuallyExclusiveFlags(ctx, "token", "san")
-		}
 	}
 
-	req, pk, err := flow.CreateSignRequest(token, sans)
+	req, pk, err := flow.CreateSignRequest(tok, sans)
 	if err != nil {
 		return err
 	}
 
-	if isStepToken {
-		// Validate that subject matches the CSR common name.
+	jwt, err := token.ParseInsecure(tok)
+	if err != nil {
+		return err
+	}
+
+	switch jwt.Payload.Type() {
+	case token.JWK: // Validate that subject matches the CSR common name.
+		if ctx.String("token") != "" && len(sans) > 0 {
+			return errs.MutuallyExclusiveFlags(ctx, "token", "san")
+		}
 		if strings.ToLower(subject) != strings.ToLower(req.CsrPEM.Subject.CommonName) {
 			return errors.Errorf("token subject '%s' and common name '%s' do not match", req.CsrPEM.Subject.CommonName, subject)
 		}
-	} else {
-		// Validate that the subject matches an email SAN
+	case token.OIDC: // Validate that the subject matches an email SAN
 		if len(req.CsrPEM.EmailAddresses) == 0 {
 			return errors.New("unexpected token: payload does not contain an email claim")
 		}
 		if email := req.CsrPEM.EmailAddresses[0]; email != subject {
 			return errors.Errorf("token email '%s' and argument '%s' do not match", email, subject)
 		}
+	case token.AWS: // Validate that the subject matches the instance id
+		if strings.ToLower(subject) != strings.ToLower(req.CsrPEM.Subject.CommonName) {
```

Comment:
I think the error messages here are confusing (not a result of this commit, because I don't think you changed the wording). Saying that the token subject doesn't match the csr is misleading. The CSR is generated from the token. So what's actually happening is that the `subject` input from the cli does not match the identity derived from the token. 

The current phrasing is confusing to me and I think we should change for all the cases. We don't need to check that the token matches the CSR b/c authority.AuthorizeSign will do that for us.

Reply:
I'll clarify a little bit these errors, those errors won't be seen by the user unless we introduce a bug when we generate the CSR

Reply:
Yea, clarification will help, especially for us when debugging.